src/object/cli_obj.c

```diff
@@ -1310,148 +1265,133 @@ obj_iod_valid(unsigned int nr, daos_iod_t *iods, bool update)
 	return true;
 }
 
-int
-dc_obj_fetch(tse_task_t *task)
+/* check if the obj request is valid */
+static int
+obj_req_valid(void *args, int opc, daos_epoch_t *epoch)
 {
-	daos_obj_fetch_t	*args = dc_task_get_args(task);
-	struct obj_auxi_args	*obj_auxi;
-	struct dc_object	*obj;
-	struct dc_obj_shard	*obj_shard;
-	int			 shard;
-	unsigned int		 map_ver;
-	uint64_t		 dkey_hash;
-	daos_epoch_t		 epoch;
-	int			 rc;
-
-	if (args->dkey == NULL || args->dkey->iov_buf == NULL || args->nr == 0
-	    || !obj_iod_valid(args->nr, args->iods, false))
-		D_GOTO(out_task, rc = -DER_INVAL);
+	daos_obj_fetch_t	*f_args;
+	daos_obj_update_t	*u_args;
+	daos_obj_punch_t	*p_args;
+	int			 rc = 0;
 
-	rc = dc_tx_check(args->th, false, &epoch);
-	if (rc) {
-		if (rc != -DER_INVAL)
-			goto out_task;
-		/* FIXME: until distributed transaction. */
-		epoch = DAOS_EPOCH_MAX; /* = daos_ts2epoch();*/
-		D_DEBUG(DB_IO, "set epoch "DF_U64"\n", epoch);
-	}
-	D_ASSERT(epoch);
+	switch (opc) {
+	case DAOS_OBJ_RPC_FETCH:
+		f_args = args;
+		if (f_args->dkey == NULL || f_args->dkey->iov_buf == NULL ||
+		    f_args->nr == 0 ||
+		    !obj_iod_valid(f_args->nr, f_args->iods, false))
+			D_GOTO(out, rc = -DER_INVAL);
+
+		rc = dc_tx_check(f_args->th, false, epoch);
+		if (rc) {
+			if (rc != -DER_INVAL)
+				D_GOTO(out, rc);
+			/* FIXME: until distributed transaction. */
+			*epoch = srv_io_dispatch ? DAOS_EPOCH_MAX :
+						   daos_ts2epoch();
```

Comment:
this can be a inline function dc_io_epoch()

Reply:
agreed.